packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/entity/mod.rs

```diff
@@ -1,27 +1,180 @@
 mod read;
 mod resolve;
 
+use std::{future::Future, pin::Pin};
+
 use async_trait::async_trait;
 use error_stack::{IntoReport, Report, Result, ResultExt};
-use futures::{stream, StreamExt, TryStreamExt};
+use futures::{stream, FutureExt, StreamExt, TryStreamExt};
 use tokio_postgres::GenericClient;
-use type_system::uri::VersionedUri;
+use type_system::uri::{BaseUri, VersionedUri};
 use uuid::Uuid;
 
 use crate::{
     knowledge::{
         Entity, EntityId, EntityQuery, EntityRootedSubgraph, PersistedEntity,
         PersistedEntityIdentifier,
     },
-    ontology::AccountId,
+    ontology::{
+        AccountId, PersistedDataType, PersistedEntityType, PersistedLinkType,
+        PersistedPropertyType, QueryDepth,
+    },
     store::{
         crud::Read,
         error::EntityDoesNotExist,
-        postgres::{ontology::EntityTypeDependencyContext, DependencyMap},
+        postgres::{
+            context::PostgresContext,
+            ontology::{EntityTypeDependencyContext, LinkTypeDependencyContext},
+            parameter_list, DependencyMap,
+        },
         AsClient, EntityStore, InsertionError, PostgresStore, QueryError, UpdateError,
     },
 };
 
+pub struct EntityDependencyContext<'a> {
+    pub referenced_data_types: &'a mut DependencyMap<VersionedUri, PersistedDataType>,
+    pub referenced_property_types: &'a mut DependencyMap<VersionedUri, PersistedPropertyType>,
+    pub referenced_link_types: &'a mut DependencyMap<VersionedUri, PersistedLinkType>,
+    pub referenced_entity_types: &'a mut DependencyMap<VersionedUri, PersistedEntityType>,
+    pub linked_entities: &'a mut DependencyMap<EntityId, PersistedEntity>,
+    pub data_type_query_depth: QueryDepth,
+    pub property_type_query_depth: QueryDepth,
+    pub link_type_query_depth: QueryDepth,
+    pub entity_type_query_depth: QueryDepth,
+    pub linked_entity_query_depth: QueryDepth,
+}
+
+impl<C: AsClient> PostgresStore<C> {
+    /// Returns the linked [`EntityId`]s and the corresponding [`LinkType`] with the provided
+    /// `entity_id` as source entity.
+    ///
+    /// [`LinkType`]: type_system::LinkType
+    async fn get_linked_entities(
```

Comment:
If you have
`EntityA` and `EntityB`

And `EntityA` --SomeLink--> `EntityB`
And `EntityB` --SomeLink--> `EntityA`

What happens in the return value?

I am starting to think more and more that we should just go to a (Vertex Set, Edge Set) representation for our return values, (for the types, entities, etc.) to avoid duplication, and to avoid worrying as much about circular dependencies and such.

Reply:
Both, `EntityA` and `EntityB` will be returned in the `linkedEntities` vector. `SomeLink` will be returned in `referencedLinkTypes` (we may want to return the link in the future as well)

Reply:
Can you give an example of a return value for something with a depth higher than 1

Reply:
Sure!
`Alice` --`FriendOf`-> `Bob`
`Bob` --`FriendOf`-> `Alice`

```http
{
  "query": {
    "eq": [
      {
        "path": [
          "version"
        ]
      },
      {
        "literal": "latest"
      }
    ]
  },
  "dataTypeQueryDepth": 0,
  "propertyTypeQueryDepth": 0,
  "linkTypeQueryDepth": 255,
  "entityTypeQueryDepth": 0,
  "linkedEntityQueryDepth": 255
}
```

```json
[
  {
    "entity": {
      "inner": {
        "http://localhost:3000/@alice/types/property-type/name/": "Alice Allison"
      },
      "identifier": {
        "entityId": "0335be0b-26a1-475f-81de-aabcb3bf0676",
        "version": "2022-09-16T10:21:38.140412Z",
        "ownedById": "74014335-79d2-4a66-8533-5a9165538497"
      },
      "typeVersionedUri": "http://localhost:3000/@alice/types/entity-type/person/v/2"
    },
    "referencedDataTypes": [],
    "referencedPropertyTypes": [],
    "referencedLinkTypes": [
      {
        "inner": {
          "$id": "http://localhost:3000/@alice/types/link-type/friend-of/v/1",
          "description": "Someone who has a shared bond of mutual affection",
          "kind": "linkType",
          "pluralTitle": "Friends Of",
          "title": "Friend Of"
        },
        "identifier": {
          "uri": "http://localhost:3000/@alice/types/link-type/friend-of/v/1",
          "ownedById": "74014335-79d2-4a66-8533-5a9165538497"
        }
      }
    ],
    "referencedEntityTypes": [],
    "linkedEntities": [
      {
        "inner": {
          "http://localhost:3000/@alice/types/property-type/name/": "Bob"
        },
        "identifier": {
          "entityId": "c0950d7e-68ee-41c8-a9bd-e1e2dddd5a2d",
          "version": "2022-09-16T10:21:38.254749Z",
          "ownedById": "74014335-79d2-4a66-8533-5a9165538497"
        },
        "typeVersionedUri": "http://localhost:3000/@alice/types/entity-type/person/v/1"
      },
      {
        "inner": {
          "http://localhost:3000/@alice/types/property-type/name/": "Alice Allison"
        },
        "identifier": {
          "entityId": "0335be0b-26a1-475f-81de-aabcb3bf0676",
          "version": "2022-09-16T10:21:38.140412Z",
          "ownedById": "74014335-79d2-4a66-8533-5a9165538497"
        },
        "typeVersionedUri": "http://localhost:3000/@alice/types/entity-type/person/v/2"
      }
    ]
  },
  {
    "entity": {
      "inner": {
        "http://localhost:3000/@alice/types/property-type/name/": "Bob"
      },
      "identifier": {
        "entityId": "c0950d7e-68ee-41c8-a9bd-e1e2dddd5a2d",
        "version": "2022-09-16T10:21:38.254749Z",
        "ownedById": "74014335-79d2-4a66-8533-5a9165538497"
      },
      "typeVersionedUri": "http://localhost:3000/@alice/types/entity-type/person/v/1"
    },
    "referencedDataTypes": [],
    "referencedPropertyTypes": [],
    "referencedLinkTypes": [
      {
        "inner": {
          "$id": "http://localhost:3000/@alice/types/link-type/friend-of/v/1",
          "description": "Someone who has a shared bond of mutual affection",
          "kind": "linkType",
          "pluralTitle": "Friends Of",
          "title": "Friend Of"
        },
        "identifier": {
          "uri": "http://localhost:3000/@alice/types/link-type/friend-of/v/1",
          "ownedById": "74014335-79d2-4a66-8533-5a9165538497"
        }
      }
    ],
    "referencedEntityTypes": [],
    "linkedEntities": [
      {
        "inner": {
          "http://localhost:3000/@alice/types/property-type/name/": "Alice Allison"
        },
        "identifier": {
          "entityId": "0335be0b-26a1-475f-81de-aabcb3bf0676",
          "version": "2022-09-16T10:21:38.140412Z",
          "ownedById": "74014335-79d2-4a66-8533-5a9165538497"
        },
        "typeVersionedUri": "http://localhost:3000/@alice/types/entity-type/person/v/2"
      },
      {
        "inner": {
          "http://localhost:3000/@alice/types/property-type/name/": "Bob"
        },
        "identifier": {
          "entityId": "c0950d7e-68ee-41c8-a9bd-e1e2dddd5a2d",
          "version": "2022-09-16T10:21:38.254749Z",
          "ownedById": "74014335-79d2-4a66-8533-5a9165538497"
        },
        "typeVersionedUri": "http://localhost:3000/@alice/types/entity-type/person/v/1"
      }
    ]
  }
]
```

Reply:
Interesting, I was wondering if you'd see the root within the linkedEntities. For now perhaps we want to just not include it, but in the long run I think we need to move to a vertex-set and edge-set representation. Although with all of our different kinds of elements it'd be pretty messy, especially if we need to still identify the root, something like this (🤮): 

```json5
[
    {
        knowledge: {
            root: id,
            entities: [

            ],
            links: {
                entityId1: entityId2
            }
        },
         ontology: {
            root: id,
            entity_types: [

            ],
            property_types: [

            ],
            data_types: [

            ],
            link_types: [

            ],
            references: {
                age: [
                    number
                ]
            }
         },
    },
]
```

Reply:
I removed the root in b4cde04292928b9ce8458ac9c1c46e5c655ba31f, this is the new output: 

```json
[
  {
    "entity": {
      "inner": {
        "http://localhost:3000/@alice/types/property-type/name/": "Alice Allison"
      },
      "identifier": {
        "entityId": "0335be0b-26a1-475f-81de-aabcb3bf0676",
        "version": "2022-09-16T10:21:38.140412Z",
        "ownedById": "74014335-79d2-4a66-8533-5a9165538497"
      },
      "typeVersionedUri": "http://localhost:3000/@alice/types/entity-type/person/v/2"
    },
    "referencedDataTypes": [],
    "referencedPropertyTypes": [],
    "referencedLinkTypes": [
      {
        "inner": {
          "$id": "http://localhost:3000/@alice/types/link-type/friend-of/v/1",
          "description": "Someone who has a shared bond of mutual affection",
          "kind": "linkType",
          "pluralTitle": "Friends Of",
          "title": "Friend Of"
        },
        "identifier": {
          "uri": "http://localhost:3000/@alice/types/link-type/friend-of/v/1",
          "ownedById": "74014335-79d2-4a66-8533-5a9165538497"
        }
      }
    ],
    "referencedEntityTypes": [],
    "linkedEntities": [
      {
        "inner": {
          "http://localhost:3000/@alice/types/property-type/name/": "Bob"
        },
        "identifier": {
          "entityId": "c0950d7e-68ee-41c8-a9bd-e1e2dddd5a2d",
          "version": "2022-09-16T10:21:38.254749Z",
          "ownedById": "74014335-79d2-4a66-8533-5a9165538497"
        },
        "typeVersionedUri": "http://localhost:3000/@alice/types/entity-type/person/v/1"
      }
    ]
  },
  {
    "entity": {
      "inner": {
        "http://localhost:3000/@alice/types/property-type/name/": "Bob"
      },
      "identifier": {
        "entityId": "c0950d7e-68ee-41c8-a9bd-e1e2dddd5a2d",
        "version": "2022-09-16T10:21:38.254749Z",
        "ownedById": "74014335-79d2-4a66-8533-5a9165538497"
      },
      "typeVersionedUri": "http://localhost:3000/@alice/types/entity-type/person/v/1"
    },
    "referencedDataTypes": [],
    "referencedPropertyTypes": [],
    "referencedLinkTypes": [
      {
        "inner": {
          "$id": "http://localhost:3000/@alice/types/link-type/friend-of/v/1",
          "description": "Someone who has a shared bond of mutual affection",
          "kind": "linkType",
          "pluralTitle": "Friends Of",
          "title": "Friend Of"
        },
        "identifier": {
          "uri": "http://localhost:3000/@alice/types/link-type/friend-of/v/1",
          "ownedById": "74014335-79d2-4a66-8533-5a9165538497"
        }
      }
    ],
    "referencedEntityTypes": [],
    "linkedEntities": [
      {
        "inner": {
          "http://localhost:3000/@alice/types/property-type/name/": "Alice Allison"
        },
        "identifier": {
          "entityId": "0335be0b-26a1-475f-81de-aabcb3bf0676",
          "version": "2022-09-16T10:21:38.140412Z",
          "ownedById": "74014335-79d2-4a66-8533-5a9165538497"
        },
        "typeVersionedUri": "http://localhost:3000/@alice/types/entity-type/person/v/2"
      }
    ]
  }
]
```